File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    queue<TreeNode*> q;
    void flatten(TreeNode* root) {
        preorder(root);
    }
    
    void preorder(TreeNode* node){
        if(!node) return ; 
        q.push(node);
        TreeNode* left = node->left;
        node->left = NULL; 
        TreeNode* right = node->right;
        node->right = NULL;
        if(q.size()>1){
            TreeNode* top = q.front();
            q.pop();
            top->right = q.front();
        }
        if(left) preorder(left);
        if(right) preorder(right);
    }
        

};

File: 0114-flatten-binary-tree-to-linked-list/0114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    def __init__(self): 
        self.prev=None
    
    def dfs(self,node): 
        if node is None: 
            return 
        
        rst = node.right
        self.prev.right=node
        self.prev=node 
        self.dfs(node.left)
        node.left=None
        self.dfs(rst) 
        

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        dummy = TreeNode(-1) 
        self.prev = dummy 
        self.dfs(root)
        return dummy.right
        
        